stl/inc/flat_map

```diff
@@ -0,0 +1,953 @@
+
+    static_assert(random_access_iterator<iterator>);
+    static_assert(convertible_to<iterator, const_iterator>);
+
+    using value_compare = _Flat_value_compare<key_type, mapped_type, key_compare>::value_compare;
+    using containers    = _Flat_Container<key_container_type, mapped_container_type>::container;
+
+public:
+    // [flat.map.cons] Constructors
+    explicit flat_map(const key_compare& _Comp) : _Key_compare(_Comp), _Data() {}
+    flat_map() : flat_map(key_compare()) {}
+
+    template <_Valid_Allocator_for_flat_map<key_container_type, mapped_container_type> Allocator>
+    explicit flat_map(const Allocator& _Alloc) : flat_map(key_compare(), _Alloc) {}
+
+    template <_Valid_Allocator_for_flat_map<key_container_type, mapped_container_type> Allocator>
+    explicit flat_map(const key_compare& _Comp, const Allocator& _Alloc)
+        : _Key_compare(_Comp), _Data{.keys = _STD make_obj_using_allocator<key_container_type>(_Alloc),
+                                   .values = _STD make_obj_using_allocator<mapped_container_type>(_Alloc)} {}
+
+    flat_map(key_container_type _Key_cont, mapped_container_type _Mapped_cont)
+        : flat_map(sorted_unique_t(), _Key_cont, _Mapped_cont) {
+        _SortAndDedup();
+    }
+
+    template <_Valid_Allocator_for_flat_map<key_container_type, mapped_container_type> Allocator>
+    flat_map(const key_container_type& _Key_cont, const mapped_container_type& _Mapped_cont, const Allocator& _Alloc)
+        : flat_map(sorted_unique_t(), _Key_cont, _Mapped_cont, _Alloc) {
+        _SortAndDedup();
+    }
+
+    flat_map(sorted_unique_t, key_container_type _Key_cont, mapped_container_type _Mapped_cont)
+        : _Key_compare(key_compare()), _Data{.keys = _STD move(_Key_cont), .values = _STD move(_Mapped_cont)} {}
+
+    template <_Valid_Allocator_for_flat_map<key_container_type, mapped_container_type> Allocator>
+    flat_map(sorted_unique_t, const key_container_type& _Key_cont, const mapped_container_type& _Mapped_cont,
+        const Allocator& _Alloc)
+        : _Key_compare(key_compare()),
+          _Data{.keys = _STD make_obj_using_allocator<key_container_type>(_Alloc, _Key_cont),
+              .values = _STD make_obj_using_allocator<mapped_container_type>(_Alloc, _Mapped_cont)} {}
+
+    template <class _InputIterator>
+        requires _Is_iterator_v<_InputIterator>
+    flat_map(_InputIterator _First, _InputIterator _Last, const key_compare& _Comp = key_compare()) : flat_map(_Comp) {
+        insert(_First, _Last);
+    }
+
+    template <class _InputIterator, _Valid_Allocator_for_flat_map<key_container_type, mapped_container_type> Allocator>
+        requires _Is_iterator_v<_InputIterator>
+    flat_map(_InputIterator _First, _InputIterator _Last, const key_compare& _Comp, const Allocator& _Alloc)
+        : flat_map(_Comp, _Alloc) {
+        insert(_First, _Last);
+    }
+
+    template <_Container_compatible_range<value_type> R>

```

Comment:
By looking at other C++23 headers, perhaps we might need to add _HAS_CXX23 guards at the top of this file to avoid the possible test failure like this :  https://github.com/microsoft/STL/pull/3817#discussion_r1237419935

Reply:
Added.